### PR TITLE
Add debug for GC

### PR DESCRIPTION
**Pull request by blackheaven**

### CODE REVIEW

**Comment by remk** [comment]
```
[bdea1c600f87187092584c9b341c1039223d715c] 10:10:25.963 [ERROR] o.a.j.t.SerialTaskManagerWorker - Task was partially performed. Check logs for more details. Taskid : TaskId{value=3f0fa459-64b0-4621-aadc-e6d5095d0eb6}
[bdea1c600f87187092584c9b341c1039223d715c] [INFO] Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.337 s - in org.apache.james.task.SerialTaskManagerWorkerTest
[bdea1c600f87187092584c9b341c1039223d715c] [INFO] Running org.apache.james.task.TaskWithIdTest
[bdea1c600f87187092584c9b341c1039223d715c] [INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0 s - in org.apache.james.task.TaskWithIdTest
[bdea1c600f87187092584c9b341c1039223d715c] [INFO] 
[bdea1c600f87187092584c9b341c1039223d715c] [INFO] Results:
[bdea1c600f87187092584c9b341c1039223d715c] [INFO] 
[bdea1c600f87187092584c9b341c1039223d715c] [ERROR] Failures: 
[bdea1c600f87187092584c9b341c1039223d715c] [ERROR]   MemoryTerminationSubscriberTest.dynamicListeningEventsShouldGetOnlyNewEvents 
[bdea1c600f87187092584c9b341c1039223d715c] Expecting:
[bdea1c600f87187092584c9b341c1039223d715c]   <[Cancelled(TaskAggregateId(TaskId{value=ff7db55c-b2fd-4d70-a02b-9cc733433398}),EventId(42),None)]>
[bdea1c600f87187092584c9b341c1039223d715c] to contain exactly (and in same order):
[bdea1c600f87187092584c9b341c1039223d715c]   <[Failed(TaskAggregateId(TaskId{value=545daac4-612e-4462-9fcc-40b202589967}),EventId(42),None,None,None),
[bdea1c600f87187092584c9b341c1039223d715c]     Cancelled(TaskAggregateId(TaskId{value=ff7db55c-b2fd-4d70-a02b-9cc733433398}),EventId(42),None)]>
[bdea1c600f87187092584c9b341c1039223d715c] but could not find the following elements:
[bdea1c600f87187092584c9b341c1039223d715c]   <[Failed(TaskAggregateId(TaskId{value=545daac4-612e-4462-9fcc-40b202589967}),EventId(42),None,None,None)]>
```

test this please

File: server/blob/blob-deduplicating/pom.xml [hunk 1]
```diff
@@ -66,6 +66,11 @@
             <artifactId>testing-base</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.typesafe.play</groupId>
+            <artifactId>play-json_${scala.base}</artifactId>
+            <version>2.8.1</version>
```
Comment (chibenwa) [inline server/blob/blob-deduplicating/pom.xml:72]: We should use the same version than in `mailbox/event/json` and JMAP RFC-8621 as far as I understand

File: server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/GCJsonReporter.scala [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+import play.api.libs.json.{JsString, Json, Writes}
+
+import scala.collection.immutable.TreeSet
+
+
+sealed trait RelatedAction
+object RelatedAction {
+  case object Init extends RelatedAction
+  case class Save(blobId: BlobId, reference: ExternalID) extends RelatedAction
+  case class Delete(reference: ExternalID) extends RelatedAction
```
Comment (chibenwa) [inline server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/GCJsonReporter.scala:32]: Delete? Or De-refenrence?

Comment (remk) [reply server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/GCJsonReporter.scala:32]: Should we update the json and the debug tool too ?

Comment (chibenwa) [reply server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/GCJsonReporter.scala:32]: I think so

File: server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/GCJsonReporter.scala [hunk 3]
```diff
@@ -44,14 +44,14 @@ object JsonReport {
...
-                   deletions: Seq[Dereference])
+                   dereferences: Seq[Dereference])
 
 
   //action
   implicit val relatedActionWrites: Writes[RelatedAction] = {
     case Init => JsString("init")
     case Save(blobId, reference) => JsString(s"save(blob = ${blobId.asString()}, reference = ${reference.id})")
-    case Delete(reference) => JsString(s"delete(reference = ${reference.id})")
+    case DereferenceAction(reference) => JsString(s"dereference(reference = ${reference.id})")
```
Comment (chibenwa) [inline server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/GCJsonReporter.scala:54]: Why add an Action suffix here and not above?

Comment (remk) [reply server/blob/blob-deduplicating/src/main/scala/org/apache/james/server/blob/deduplication/GCJsonReporter.scala:54]: It was to disembiguate between the different `Dereference` class but i think i will just prefix with the RelatedAction object.

File: server/blob/blob-deduplicating/src/test/scala/org/apache/james/server/blob/deduplication/GCJsonReporterTest.scala [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            firstDeleteReport,
+            JsonReport.State(GarbageCollect,
+              `reference-generations` = TreeSet(generation),
+              `garbage-collection-iterations` = TreeSet(initialIteration, firstIteration),
+              blobs = Nil,
+              references = Nil,
+              deletions = Nil )))
+        }
+
+        "one reference is added, a gc run two generation later, then  it is removed and the GC is ran again" in {
```
Comment (rouazana) [inline server/blob/blob-deduplicating/src/test/scala/org/apache/james/server/blob/deduplication/GCJsonReporterTest.scala:126]:
```suggestion
        "one reference is added, a gc run two generations later, then it is removed and the GC is run again" in {
```

File: server/blob/blob-deduplicating/src/test/scala/org/apache/james/server/blob/deduplication/GCJsonReporterTest.scala [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            `garbage-collection-iterations` = TreeSet(initialIteration, firstIteration),
+            blobs = Seq[JsonReport.BlobId](JsonReport.BlobId(blobId.asString, blobId.generation)),
+            references = Seq(JsonReport.Reference(externalId.id, blobId.asString, generation)),
+            deletions = Seq(JsonReport.Dereference(blobId.asString(), generation, initialIteration)))))
+      }
+    }
+
+    "GC has been ran" when {
+      "report added and removed references" when {
+        "one reference is added then removed and the GC is ran 2 generation later" in {
```
Comment (rouazana) [inline server/blob/blob-deduplicating/src/test/scala/org/apache/james/server/blob/deduplication/GCJsonReporterTest.scala:107]:
```suggestion
        "one reference is added then removed and the GC is run 2 generations later" in {
```

File: server/blob/blob-deduplicating/src/test/scala/org/apache/james/server/blob/deduplication/GCJsonReporterTest.scala [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+    val iterationForImmediateGC = Iteration(1L, Set(), generation)
+    val gcReportImmediate = GCIterationReport(iterationForImmediateGC, Set())
+
+    "be minimal" when {
+      "on initial state" in {
+        GCJsonReporter
+          .report(GCIterationEvent(gcReportImmediate) :: Nil)
+          .states should be (Seq(initialReport,
+          JsonReport.State(GarbageCollect,
```
Comment (rouazana) [inline server/blob/blob-deduplicating/src/test/scala/org/apache/james/server/blob/deduplication/GCJsonReporterTest.scala:65]: indent (and same in other tests)

**Comment by chibenwa** [comment]
```
[bfd6eeddbfc2fd6adce91b82c34da9849a83eb45] [ERROR] Failed to execute goal com.google.cloud.tools:jib-maven-plugin:1.5.1:buildTar (docker packaging) on project mock-smtp-server: com.google.cloud.tools.jib.api.RegistryUnauthorizedException: Unauthorized for registry-1.docker.io/library/openjdk: 401 Unauthorized
[bfd6eeddbfc2fd6adce91b82c34da9849a83eb45] [ERROR] {"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":[{"Type":"repository","Class":"","Name":"library/openjdk","Action":"pull"}]}]}
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**